### DISCUSSION TITLE
IDs in addition to tags

### DISCUSSION THREAD

**Discussion by oaklandgit** [replit/kaboom#371] [Ideas]
I know that a class of game objects can be referenced by tag, as in:

```
const enemies = get("enemy")
```
and that a single game object can be referenced via variable, as in:
```
const enemy1 = add([ etc ])
```

but letting Kaboom manage unique objects would be useful, and likely easier for newbies like:

```
add([
   ID: "somethingUnique"
])
```

which could later be referenced as:

```
get("somethingUnique")
```

**Comment by lajbel** [comment]
hmmm, yeah, something like id comp, like a tag but unique

**Comment by slmjkdbtl** [comment]
Sounds like a sugar for `get("name")[0]`? A write barrier might be helpful to ensure there's no id is reused. This kind of custom component should do it:

```ts
const nameRegistry: Record<string, GameObj> = {};

function name(name: string) {
	return {
		id: "name",
		name: name,
		add() {
			if (nameRegistry[name]) {
				throw new Error(`Name already exists: ${name}`)
			}
			nameRegistry[name] = this;
		},
		destroy() {
			delete nameRegistry[name];
		},
	}
}

function getUnique(name: string) {
	return nameRegistry[name];
}

add([
    sprite("froggy"),
    name("player"),
])
```

One thing is I think `get()` should always return an array to avoid unpredictability, thats why I put the awkwardly named `getUnique()` up there

**Reply by oaklandgit** [reply]
Yes, this is exactly what I had in mind. (And I don't think getUnique() is awkward at all. Tho maybe getById() is another option.)
